base/src/main/java/io/spine/validate/AbstractValidatingBuilder.java

```diff
@@ -224,6 +226,28 @@ private void validateResult(T message) throws ValidationException {
         checkViolations(violations);
     }
 
+    @SuppressWarnings("unused")
+        // Called by all actual validating builder subclasses.
+    protected final void validateSetOnce(FieldDescriptor descriptor) throws ValidationException {
+        boolean setOnce = descriptor.getOptions()
+                                    .getExtension(OptionsProto.setOnce);
+        boolean valueAlreadySet = getMessageBuilder().hasField(descriptor);
+        if (setOnce && valueAlreadySet) {
+            throw setOnceViolation(descriptor);
+        }
+    }
+
+    private static ValidationException setOnceViolation(FieldDescriptor descriptor) {
```

Comment:
```suggestion
    private static ValidationException violatedSetOnce(FieldDescriptor descriptor) {
```

Otherwise, this method looks like a setter.